include/boost/url/impl/decoded_view.hpp

```diff
@@ -0,0 +1,323 @@
+//
+// Copyright (c) 2022 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_IMPL_PCT_DECODED_RANGE_HPP
+#define BOOST_URL_IMPL_PCT_DECODED_RANGE_HPP
+
+namespace boost {
+namespace urls {
+
+class decoded_view::const_iterator
+{
+    char const* begin_{nullptr};
+    char const* pos_{nullptr};
+    char const* end_{nullptr};
+    bool plus_to_space_{true};
+
+    friend decoded_view;
+
+    constexpr
+    const_iterator(
+        char const* str,
+        size_type n,
+        bool plus_to_space) noexcept
+        : begin_(str)
+        , pos_(str)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+    // end ctor
+    constexpr
+    const_iterator(
+        char const* str,
+        size_type n,
+        bool plus_to_space,
+        int) noexcept
+        : begin_(str)
+        , pos_(str + n)
+        , end_(str + n)
+        , plus_to_space_(plus_to_space)
+    {}
+
+public:
+    using value_type = char;
+    using reference = char;
+    using pointer = void const*;
+    using const_reference = char;
+    using size_type = std::size_t;
+    using difference_type = std::ptrdiff_t;
+    using iterator_category =
+        std::bidirectional_iterator_tag;
+
+    const_iterator() noexcept = default;
+
+    const_iterator(const_iterator const&) noexcept = default;
+
+    const_iterator&
+    operator=(const_iterator const&) noexcept = default;
+
+    reference
+    operator*() const noexcept
+    {
+        char c{'\0'};
+        pct_decode_opts opt;
+        opt.plus_to_space = plus_to_space_;
+        pct_decode_unchecked(
+            &c, &c+1, string_view(pos_, end_), opt);
+        return c;
+    }
+
+    const_iterator&
+    operator++() noexcept
+    {
+        BOOST_ASSERT(pos_ != nullptr);
+        pos_ += *pos_ == '%' ? 3 : 1;
+        return *this;
+    }
+
+    const_iterator&
+    operator--() noexcept
+    {
+        BOOST_ASSERT(pos_ != begin_);
+        if (pos_ - begin_ >= 3 &&
+            *(pos_ - 3) == '%')
+            pos_ -= 3;
+        else
+            --pos_;
+        return *this;
+    }
+
+    const_iterator
+    operator++(int) noexcept
+    {
+        auto tmp = *this;
+        ++*this;
+        return tmp;
+    }
+
+    const_iterator
+    operator--(int) noexcept
+    {
+        auto tmp = *this;
+        --*this;
+        return tmp;
+    }
+
+    const char*
+    base()
```

Comment:
What's this?

Reply:
All view range iterators have a `base()` function to get the underlying iterator.

Example: https://en.cppreference.com/w/cpp/ranges/transform_view/iterator

Reply:
But this is no different from calling `encoded().data()`.  What is the use case here? `transform_view` is a class template which is used in generic contexts. We should not just add things because something else has it. Leave it out and wait until a user requests it...

Reply:
The use case is conforming to the standard so that STL algorithms work on the range. 

Yes, I know: "People on the committee are not smart", etc... and we should do things differently. 
But it's not in my control that other users will expect us to follow their standards and not expect them to follow our standards.

![](https://imgs.xkcd.com/comics/standards.png)

In any case, keeping this is not something I'm very attached to. I'm not emotionally attached to anything. This is just the rationale.